### PR TITLE
Properly rewind chain and remove some weirdness in callbacks.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libethereum/Client.cpp [hunk 1]
```diff
@@ -787,6 +788,16 @@ bool Client::submitSealed(bytes const& _header)
 void Client::rewind(unsigned _n)
 {
 	bc().rewind(_n);
+	onChainChanged(ImportRoute{});
```
Comment (winsvega) [inline libethereum/Client.cpp]: /libethereum/libethereum/Client.cpp:791: error: missing initializer for member 'dev::eth::ImportRoute::deadBlocks' [-Werror=missing-field-initializers]
  onChainChanged(ImportRoute{});
                             ^

**Comment by chriseth** [comment]
Working properly now.

File: libethereum/BlockChain.h [hunk 2]
```diff
@@ -350,6 +349,11 @@ class BlockChain
 
 	void checkConsistency();
 
+	/// Clears all caches from the tip of the chain up to (including) _firstInvalid.
+	/// These incude the blooms, the block hashes and the transaction lookup tables.
```
Comment (bobsummerwill) [inline libethereum/BlockChain.h]: typo in comment ... should be include

**Comment by LefterisJP** [comment]
This PR is awesome.

**Comment by winsvega** [comment]
working on this. looks good

**Comment by smartbitcoin** [comment]
so glad to see this patch,   I'll test it  to confirm that it 'll fix the "pure eth++ private net will stuck randomly bug" .

**Pull request merged**